### PR TITLE
Check access token for expiration before API calls

### PR DESCRIPTION
**Pull request by daybreaker**
connects #342 

Adds a refreshCheck promise that can be called before an API call, that will check if the access token is expired, and use the refresh token to renew it if it is.

Also wrapped more api calls in try/catch to prevent freezing the app on errors

Also refactored a bunch of server code to reduce duplicate code, and move other code in to more reasonable locations.

### CODE REVIEW

File: src/constants/Server.js [hunk 1]
```diff
@@ -29,3 +28,27 @@ export function apiErrorAlert(error, defaultMsg = 'There was a problem.') {
...
+
+function parseJwt(token) {
+  const base64Url = token.split('.')[1];
+  const base64 = base64Url.replace('-', '+').replace('_', '/');
+
+  return JSON.parse(window.atob(base64));
+}
+
+export async function refreshCheck() {
+  /* eslint-disable complexity,camelcase */
```
Comment (jwietelmann) [inline src/constants/Server.js]: Don't disable complexity.

File: src/state/authStateProvider.js [hunk 2]
```diff
@@ -108,6 +109,7 @@ class AuthContainer extends Container {
     }
   }
 
+  /* eslint-disable complexity */
```
Comment (jwietelmann) [inline src/state/authStateProvider.js]: Don't disable complexity.

File: src/state/eventManagerStateProvider.js [hunk 3]
```diff
@@ -1,9 +1,9 @@
 import {Container} from 'unstated'
-import {server} from '../constants/Server'
+import {server, apiErrorAlert, refreshCheck} from '../constants/Server'
 
 const SCAN_MESSAGE_TIMEOUT = 3000;
 
-/* eslint-disable camelcase */
+/* eslint-disable camelcase,complexity,space-before-function-paren */
```
Comment (jwietelmann) [inline src/state/eventManagerStateProvider.js]: Don't disable complexity.

File: src/constants/Server.js [hunk 4]
```diff
@@ -11,9 +12,7 @@ const authString = basicAuthUsername || basicAuthPassword ? `${basicAuthUsername
 
 export const BASE_URL = `https://${authString}staging.bigneon.com`;
 
-export const server = new Bigneon.Server({prefix: `${BASE_URL}/api`})// , {}, mocker)
-
-// eslint-disable-next-line complexity
+/* eslint-disable complexity */
```
Comment (jwietelmann) [inline src/constants/Server.js]: Don't disable complexity.

**Comment by daybreaker** [comment]
@jwietelmann hoping this looks better

File: src/state/cartStateProvider.js [hunk 5]
```diff
@@ -57,6 +57,7 @@ class CartContainer extends Container {
     }]
 
     try {
+      await refreshCheck()
```
Comment (jwietelmann) [inline src/state/cartStateProvider.js]: Making the caller do this manually is kind of a time-bomb. I have a far-out suggestion for getting rid of every single one of these calls, and it looks a little like this: https://codepen.io/anon/pen/REWgOo?editors=0012

**Pull request merged**